userspace/engine/falco_engine.cpp

```diff
@@ -427,27 +431,288 @@ std::size_t falco_engine::add_source(const std::string &source,
+		{
+			exception_operators.append(e.comps.item);
+		}	
+	}
+	rule["details"]["exceptionOperators"] = exception_operators;
+
+	if(ri.source == falco_common::syscall_source)
+	{
+		// Store event types
+		Json::Value events;
+		get_json_evt_types(ast.get(), events);
+		rule["details"]["events"] = events;
+	}
+}
+
+void falco_engine::get_json_details(const rule_loader::macro_info& m,
+	Json::Value& macro) const
+{
+	Json::Value macro_info;
+
+	macro_info["name"] = m.name;
+	macro_info["condition"] = m.cond;
+	macro["info"] = macro_info;
+
+	// Assumption: no exception because rules have already been loaded.
+	auto ast = libsinsp::filter::parser(m.cond).parse();
+
+	Json::Value json_details;
+	get_json_details(ast.get(), json_details);
+	macro["details"] = json_details;
+
+	// Store event types
+	Json::Value events;
+	get_json_evt_types(ast.get(), events);
+	macro["details"]["events"] = events;
+}
+
+void falco_engine::get_json_details(const rule_loader::list_info& l, 
+	Json::Value& list) const
+{
+	Json::Value list_info;
+	list_info["name"] = l.name;
+	list["info"] = list_info;
+
+	Json::Value items = Json::arrayValue;
+	Json::Value lists = Json::arrayValue;
+	for(const auto &i : l.items)
+	{
+		if(m_rule_collector.lists().at(i) != nullptr)
+		{
+			lists.append(i);
+			continue;
+		}
+		items.append(i);
+	}
+
+	list["details"]["items"] = items;
+	list["details"]["lists"] = lists;
+}
+
+void falco_engine::get_json_details(libsinsp::filter::ast::expr* ast,
+	Json::Value& output) const
+{
+	filter_details details;
+	for(const auto &m : m_rule_collector.macros())
+	{
+		details.known_macros.insert(m.name);
+	}
+
+	for(const auto &l : m_rule_collector.lists())
+	{
+		details.known_lists.insert(l.name);
+	}
+
+	// Resolve the AST details
+	filter_details_resolver resolver;
+	resolver.run(ast, details);
+
+	Json::Value macros = Json::arrayValue;
+	for(const auto &m : details.macros)
+	{
+		macros.append(m);
+	}
+	output["macros"] = macros;
+
+	Json::Value operators = Json::arrayValue;
+	for(const auto &o : details.operators)
+	{
+		operators.append(o);
+	}
+	output["operators"] = operators;
+
+	Json::Value condition_fields = Json::arrayValue;
+	for(const auto &f : details.fields)
+	{
+		condition_fields.append(f);
+	}
+	output["conditionFields"] = condition_fields;
+
+	Json::Value lists = Json::arrayValue;
+	for(const auto &l : details.lists)
+	{
+		lists.append(l);
+	}
+	output["lists"] = lists;
+	
+	details.reset();
+}
+
+void falco_engine::get_json_evt_types(libsinsp::filter::ast::expr* ast,
+					Json::Value& output) const
+{
+	output = Json::arrayValue;
+	auto evttypes = libsinsp::filter::ast::ppm_event_codes(ast);
+	if(evttypes.size() != libsinsp::events::all_event_set().size())
+	{
+		std::unordered_set<std::string> evts;
+		for(const auto &e : evttypes)
+		{
+			auto evt_info = libsinsp::events::info(e);
+			auto res = evts.insert(std::string(evt_info->name));
+			if(res.second)
+			{
+				output.append(evt_info->name);
+			}
+		}
+	}
+}
```

Comment:
```suggestion
	auto evtcodes = libsinsp::filter::ast::ppm_event_codes(ast);
	if(evtcodes.size() != libsinsp::events::all_event_set().size())
	{
		auto syscodes = libsinsp::filter::ast::ppm_sc_codes(ast);
		for (const auto& n : libsinsp::events::sc_set_to_event_names(syscodes))
		{
			output.append(n);
		}
		for (const auto& n : libsinsp::events::event_set_to_names(evtcodes, false))
		{
			output.append(n);
		}
	}
}
```

I know it's not intuitive... but I think that's what we want (this is how evt.type works under the hood):
- names from the event table for all non-generic events
- names from the syscall table for all generic events

Reply:
Thank you, I totally missed this. Added this suggestion plus the merging of the two sets so that we don't have duplicates